signimage/signimage.go

```diff
@@ -0,0 +1,352 @@
+package main
+
+import (
+	//"context"
+	"archive/tar"
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"github.com/docker/cli/cli/config"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	//"github.com/docker/cli/cli/config/configfile"
+	dockertypes "github.com/docker/cli/cli/config/types"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+)
+
+func getenv(key string, fallback string) string {
+	value, exists := os.LookupEnv(key)
+	if !exists {
+		value = fallback
+	}
+	return value
+}
+
+func checkarg(arg *string, envvar string, fallback string) {
+	if *arg == "" {
+		if envvar != "" {
+			v, exists := os.LookupEnv(envvar)
+			if exists {
+				*arg = v
+			} else if !exists && fallback != "" {
+				*arg = fallback
+			} else {
+				fmt.Printf("%s not found:\n", envvar)
+				flag.PrintDefaults()
+				os.Exit(0)
+			}
+		}
+	}
+	return
+}
+
+func canonicalisePath(path string) string {
+	canonical := strings.Replace("/"+path, "/./", "/", -1)
+	return strings.Replace(canonical, "//", "/", -1)
+}
+
+func extractFile(tmpdir string, header *tar.Header, tarreader io.Reader) string {
+
+	contents := make([]byte, header.Size)
+	offset := 0
+	for {
+		rc, err := tarreader.Read(contents[offset:])
+		if err != nil && err != io.EOF {
+			die(2, "could not read ko file", err)
+		}
+		offset += rc
+		if err == io.EOF {
+			break
+		}
+	}
+	dirname := filepath.Dir(header.Name)
+	err := os.MkdirAll(tmpdir+"/"+dirname, 0750)
+	if err != nil {
+		die(2, "could not create tempdir for kmod", err)
```

Comment:
I would not call `os.Exit` from anywhere except the `main` function. The rest of the code should return errors to be processed by `main`.

Reply:
I'm leaving the die() helper function calls in main() because to unroll them would result in lots of repetitive logging/exit code, but you're right calling it from many layers down the stack in the image walking code is pretty nasty. Fixed.